### ISSUE TITLE
CMake compile on Linux fails (can't find installed libxml2 headers)

### ISSUE THREAD

**Issue by bacon-cheeseburger** [nzbgetcom/nzbget#85]
Hi,
Just found this repo (previously using nzbget-ng). Cloned develop branch and attempted CMake compile on Debian Linux (sid). Please let me know if more info is needed:

Installed:
libxml2:amd64                                  2.9.14+dfsg-1.3+b1
libxml2-dev:amd64                              2.9.14+dfsg-1.3+b1

$ find /usr/include/libxml* -name "parser.h" -type f
/usr/include/libxml2/libxml/parser.h

...
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.14")
...

$ grep -i libxml2 /test/nzbget-r2376.git+becca5eb/build/CMakeCache.txt
LIBXML2_INCLUDE_DIR:PATH=/usr/include/libxml2
LIBXML2_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libxml2.so
LIBXML2_XMLLINT_EXECUTABLE:FILEPATH=LIBXML2_XMLLINT_EXECUTABLE-NOTFOUND
pkgcfg_lib_PC_LIBXML_xml2:FILEPATH=/usr/lib/x86_64-linux-gnu/libxml2.so
//Details about finding LibXml2
FIND_PACKAGE_MESSAGE_DETAILS_LibXml2:INTERNAL=[/usr/lib/x86_64-linux-gnu/libxml2.so][/usr/include/libxml2][v2.9.14()]
//ADVANCED property for variable: LIBXML2_INCLUDE_DIR
LIBXML2_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: LIBXML2_LIBRARY
LIBXML2_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: LIBXML2_XMLLINT_EXECUTABLE
LIBXML2_XMLLINT_EXECUTABLE-ADVANCED:INTERNAL=1
PC_LIBXML_CFLAGS:INTERNAL=-I/usr/include/libxml2
PC_LIBXML_INCLUDE_DIRS:INTERNAL=/usr/include/libxml2
PC_LIBXML_STATIC_CFLAGS:INTERNAL=-I/usr/include/libxml2
PC_LIBXML_STATIC_INCLUDE_DIRS:INTERNAL=/usr/include/libxml2

In file included from /test/nzbget-r2376.git+becca5eb/lib/par2/datablock.cpp:20:
/test/nzbget-r2376.git+becca5eb/daemon/main/nzbget.h:226:10: fatal error: libxml/parser.h: No such file or directory
  226 | #include <libxml/parser.h>
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.

**Comment by paul-chambers** [comment]
I also encountered this. I added 'include_directories(/usr/include/libxml2) to CMakeLists.txt to work around this for now.

**Comment by bacon-cheeseburger** [comment]
@paul-chambers 
Thanks for the tip. Btw, have you been able to build a usable binary? With ENABLE_TESTS=ON, it fails to build. With ENABLE_TESTS=OFF it only builds targets Yencode and Par2 but nothing else. Would this be the result of incomplete CMake build support or??

**Comment by paul-chambers** [comment]
Yep, the CMakeLIsts.txt file only refers to the tests, there's no target provided for nzbget itself. Nor is there any wrapper/glue yet that teaches CMake how to use the existing autotools/autoconf stuff to configure & make, as a stepping stone, before making that obsolescent by converting all that messy autoconf stuff over to native cmake logic.

I have built a fresh, working nzbget executable, but only using the original autoconf/configure/make method. And even that hasn't been updated to use current versions of autotools. Something I'd already done for nzbget-ng, so I plan to port that across at a minimum.

I should be able to hide the autotools stuff behind `cmake`, in a similar fashion to the stub makefile I created for nzbget-ng. In 'short', make a target for `./configure` that depends on `configure.ac.in` to generate it, have `./Makefile` depend on `./configure`, and `./nzbget` depend on `./Makefile`. then you can build nzbget and have cmake figure out what's missing and what to do about it. It seems bizarre to be handling this manually :)  As usual, it's a matter of finding time to save time...

Call me crazy, but I like to have my projects just do the right thing when I type make (or invoke cmake). What needs to be done to create whatever's missing is more easily expressed (and less error-prone, if done well) in a makefile/CMakeLists.txt than described in build instructions to a human :)

**Comment by bacon-cheeseburger** [comment]
Ahh, ok, I didn't realize the build situation was as you described but am glad it's on your (and the other devs) radar. I'm not sure how far this fork diverges from your -ng; Hopefully the gap isn't massive enough to be problematic; Or, if it's due for an overhaul of sorts, the timing is good for that to happen. As an end-user, I'm appreciative there are guys committed to the continued development of such a useful tool.

**Comment by dnzbk** [comment]
So, the story is that we decided to switch from autotools (macOS, POSIX) + Visual Studio build system (Windows) on CMake as a more universal build system.
As a first step, we started the migration to CMake by fixing the tests on Windows, as they had't worked before.
The next step is to add a build of nzbget itself and tests for the main platforms in CMake. This has already been almost done in my [dev branch](https://github.com/nzbgetcom/nzbget/tree/feature/cmake).
And the last step will be to add embedded platforms.
